### ISSUE TITLE
a question about ddi rate

### ISSUE THREAD

**Issue by azusakou** [sunlabuiuc/PyHealth#79]
Hi! sorry for bothering you again.

I added ddi rate as a metric and checked the ddi rate, but this value was much smaller than it should be. Besides, it seems that the ddi matrix of PyHealth is different from SafeDrug and GAMEnet. Could you please give me some help about it?

**Comment by ycq091044** [comment]
Here is the ddi from my SafeDrug paper https://github.com/ycq091044/SafeDrug.

<img width="479" alt="image" src="https://user-images.githubusercontent.com/22296469/216382185-9f2fa568-2aa8-43dc-adee-f86c5aaf8510.png">

Here is the ddi definition from pyhealth:

<img width="600" alt="image" src="https://user-images.githubusercontent.com/22296469/216382361-7e930817-39db-4b00-96ca-00d1a12d29fd.png">

Both flatten the predictions into visit levels and then calculate the DDI. The function itself is the same.

**Comment by ycq091044** [comment]
Could you please show me how you extract the DDI matrix and how you use the ddi_rate_score function?

**Comment by azusakou** [comment]
> Could you please show me how you extract the DDI matrix and how you use the ddi_rate_score function?

Thanks for your reply.

Here is how I extract the DDI matrix in calculate_loss:
~~~
def calculate_loss(
        self, logits: torch.Tensor, y_prob: torch.Tensor, labels: torch.Tensor
    ) -> torch.Tensor:
        mul_pred_prob = y_prob.T @ y_prob  # (voc_size, voc_size)
        batch_ddi_loss = (
            torch.sum(mul_pred_prob.mul(self.ddi_adj)) / self.ddi_adj.shape[0] ** 2
        )

        y_pred = y_prob.detach().cpu().numpy()
        y_pred[y_pred >= 0.5] = 1
        y_pred[y_pred < 0.5] = 0
        y_pred = [np.where(sample == 1)[0] for sample in y_pred]
        '''        
        import dill
        with open('ddi.pkl', 'wb') as f:
            dill.dump(self.ddi_adj, f)
        '''
        cur_ddi_rate = ddi_rate_score(y_pred, self.ddi_adj.cpu().numpy())
        if cur_ddi_rate > self.target_ddi:
            beta = max(0.0, 1 + (self.target_ddi - cur_ddi_rate) / self.kp)
            add_loss, beta = batch_ddi_loss, beta
        else:
            add_loss, beta = 0, 1

        # obtain target, loss, prob, pred
        bce_loss = self.loss_fn(logits, labels)

        loss = beta * bce_loss + (1 - beta) * add_loss
        return loss
~~~
further, here is how I use did rate function:
~~~
def ddi_rate_score_w(record, path="./ddi.pkl"):
    # ddi rate
    ddi_A = dill.load(open(path, "rb"))
    all_cnt = 0
    dd_cnt = 0
    for patient in record:
        for adm in patient:
            med_code_set = adm
            for i, med_i in enumerate(med_code_set):
                for j, med_j in enumerate(med_code_set):
                    if j <= i:
                        continue
                    all_cnt += 1
                    if ddi_A[med_i, med_j] == 1 or ddi_A[med_j, med_i] == 1:
                        dd_cnt += 1
    if all_cnt == 0:
        return 0
    return dd_cnt / all_cnt
~~~

**Comment by ycq091044** [comment]
where did you get "./ddi.pkl"

**Comment by azusakou** [comment]
> where did you get "./ddi.pkl"

from calculate_loss, I saved self.ddi_adj as "./ddi.pkl"
~~~
def calculate_loss(
        self, logits: torch.Tensor, y_prob: torch.Tensor, labels: torch.Tensor
    ) -> torch.Tensor:
        mul_pred_prob = y_prob.T @ y_prob  # (voc_size, voc_size)
        batch_ddi_loss = (
            torch.sum(mul_pred_prob.mul(self.ddi_adj)) / self.ddi_adj.shape[0] ** 2
        )

        y_pred = y_prob.detach().cpu().numpy()
        y_pred[y_pred >= 0.5] = 1
        y_pred[y_pred < 0.5] = 0
        y_pred = [np.where(sample == 1)[0] for sample in y_pred]
        '''        

        import dill  
        with open('ddi.pkl', 'wb') as f:
            dill.dump(self.ddi_adj, f)
        '''
        cur_ddi_rate = ddi_rate_score(y_pred, self.ddi_adj.cpu().numpy())
        if cur_ddi_rate > self.target_ddi:
            beta = max(0.0, 1 + (self.target_ddi - cur_ddi_rate) / self.kp)
            add_loss, beta = batch_ddi_loss, beta
        else:
            add_loss, beta = 0, 1

        # obtain target, loss, prob, pred
        bce_loss = self.loss_fn(logits, labels)

        loss = beta * bce_loss + (1 - beta) * add_loss
        return loss
~~~

**Comment by azusakou** [comment]
I am a little confused, self.ddi_adj has different shape (191, 191) with  [ddi_A_final.pkl](https://github.com/ycq091044/SafeDrug/blob/main/data/output/ddi_A_final.pkl) (110, 110)

**Comment by ycq091044** [comment]
Thanks. Could you tell me how is your previous DDI rate and what is the current DDI rate?

We do clean up the pipeline and change some places, which makes the final drug set different (size 191) versus (size 110). I remember the main part is here:
<img width="505" alt="image" src="https://user-images.githubusercontent.com/22296469/216486135-7ff3f36e-90c4-4eef-aab9-a63b0fb0c379.png">

In Safedrug, the atc2smiles function missed some mappings, so the corresponding drugs are empty in that record. We thus directly remove the record (the remaining # drugs is 110). However, we find a better way to do the atc2smiles mapping in pyhealth.models.safedrug. With more comprehensive mapping, more drugs are left after our filtering (191 drugs), leading to different final DDI rate.

**Comment by azusakou** [comment]
> Thanks. Could you tell me how is your previous DDI rate and what is the current DDI rate?
> 
> We do clean up the pipeline and change some places, which makes the final drug set different (size 191) versus (size 110). I remember the main part is here: <img alt="image" width="505" src="https://user-images.githubusercontent.com/22296469/216486135-7ff3f36e-90c4-4eef-aab9-a63b0fb0c379.png">
> 
> In Safedrug, the atc2smiles function missed some mappings, so the corresponding drugs are empty in that record. We thus directly remove the record (the remaining # drugs is 110). However, we find a better way to do the atc2smiles mapping in pyhealth.models.safedrug. With more comprehensive mapping, more drugs are left after our filtering (191 drugs), leading to different final DDI rate.

Many thanks! I see.
The previous ddi rate was around 0.058, and the current was around 0.024
